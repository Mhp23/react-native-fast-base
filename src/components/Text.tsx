import React from 'react';
import {TextProps} from '../core/types';
import {
  StyleProp,
  Text as NativeText,
  TextStyle,
  TouchableOpacity,
} from 'react-native';
import {responsiveFontSize} from 'react-native-responsive-dimensions';
import {useTheme} from '../hooks';

const defaultProps: TextProps = {
  alignX: 'auto',
};

const Text: React.FC<TextProps> = ({
  children,
  lineHeight,
  alignX,
  style,
  color,
  font,
  size,
  onPress,
  ...rest
}) => {
  const {colors} = useTheme();

  const textStyles: StyleProp<TextStyle> = {
    fontFamily: font,
    fontSize: responsiveFontSize(size || 1.5),
    color: color ? color : colors?.text,
    textAlign: alignX,
    lineHeight: lineHeight,
  };

  return onPress ? (
    <TouchableOpacity activeOpacity={0.5} onPress={onPress}>
      <NativeText style={[textStyles, style]} {...rest}>
        {children}
      </NativeText>
    </TouchableOpacity>
  ) : (
    <NativeText style={[textStyles, style]} {...rest}>
      {children}
    </NativeText>
  );
};

Text.defaultProps = defaultProps;

export default Text;
