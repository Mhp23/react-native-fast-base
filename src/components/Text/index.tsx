import React from 'react';
import {useStyle, useTheme} from '../../hooks';
import {useRM} from 'react-native-full-responsive';
import {colorSelector, makeStyle} from '../../utils';
import {DefaultTextSizes, type TextProps} from '../../core';
import {StyleSheet, Text as NativeText, type TextStyle} from 'react-native';

const defaultProps: TextProps = {
  size: 'md',
};
/**
 * in typescript, the Text component allows you to use custom fonts that
 * you added to your project, only need to wrap your type inside it like
 * the below example:
 * ```
 * <Text<'MyFont1' | 'MyFont2'> {...props}>
 *  Awesome Text!
 * </Text>
 * ```
 */
const Text = React.forwardRef(
  <T extends string = ''>(
    {
      ax,
      font,
      size,
      color,
      style,
      weight,
      height,
      children,
      ...rest
    }: TextProps<T>,
    ref?: React.Ref<NativeText>,
  ) => {
    const {colors} = useTheme();

    const {rs} = useRM();

    const textStyles = useStyle<TextStyle>(() => {
      const textColor = colorSelector(color) || colors?.text;
      const fontSize = rs(
        typeof size === 'number' ? size : DefaultTextSizes[size],
      );
      return makeStyle({
        fontSize,
        textAlign: ax,
        fontFamily: font,
        color: textColor,
        fontWeight: weight,
        lineHeight: height,
      });
    }, [color, colors?.text, rs, size, font, ax, weight, height]);

    return (
      <NativeText
        ref={ref}
        testID="FAST_BASE_TEXT"
        style={StyleSheet.flatten([textStyles, style])}
        {...rest}>
        {children}
      </NativeText>
    );
  },
);

Text.defaultProps = defaultProps;

export default Text as <T extends string = ''>(
  props: TextProps<T> & {ref?: React.Ref<NativeText>},
) => React.ReactElement;
