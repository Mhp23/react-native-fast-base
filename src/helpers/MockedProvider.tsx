import React from 'react';
import {DefaultTheme, ThemeContext, type ThemeContextProps} from '../core';

const MockedProvider: React.FC = ({children}) => {
  return (
    <ThemeContext.Provider
      value={{
        ...({} as ThemeContextProps),
        ...DefaultTheme.DefaultTheme,
      }}>
      {children}
    </ThemeContext.Provider>
  );
};

export default MockedProvider;
