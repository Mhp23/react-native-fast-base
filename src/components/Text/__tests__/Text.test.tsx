import React from 'react';
import Text from '..';
import {rs} from 'react-native-full-responsive';
import MockedProvider from '../../../helpers/MockedProvider';
import {render, cleanup} from '@testing-library/react-native';

afterEach(cleanup);

const TextID = 'FAST_BASE_TEXT';

const defaultSize = 14;

describe('Text component', () => {
  it('should match snapshot', () => {
    const rendered = render(
      <MockedProvider>
        <Text size={defaultSize}>It's fast base text!</Text>
      </MockedProvider>,
    ).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  it('should render default font size correctly', () => {
    const calculatedFontSize = rs(defaultSize);
    const {getByTestId} = render(
      <Text size={defaultSize}>It's fast base text!</Text>,
    );
    const textComponent = getByTestId(TextID);
    expect(textComponent.props.style).toMatchObject({
      fontSize: calculatedFontSize,
    });
  });

  it('should render props correctly', () => {
    const calculatedFontSize = rs(defaultSize);
    const rendered = render(
      <MockedProvider>
        <Text<'MyFont1' | 'MyFont2' | 'MyFont3'>
          size={defaultSize}
          font="MyFont2"
          color="red">
          It's fast base text!
        </Text>
      </MockedProvider>,
    );
    expect(rendered).toBeTruthy();
    const textComponent = rendered.getByTestId(TextID);
    expect(textComponent.props.style).toMatchObject({
      color: 'red',
      fontFamily: 'MyFont2',
      fontSize: calculatedFontSize,
    });
  });
});
