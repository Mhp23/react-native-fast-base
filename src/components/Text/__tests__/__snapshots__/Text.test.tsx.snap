// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text component should match snapshot 1`] = `
<Text
  style={
    Object {
      "color": "#000000",
      "fontSize": 29.166666666666668,
    }
  }
  testID="FAST_BASE_TEXT"
>
  It's fast base text!
</Text>
`;
